File: Dong/c/CirList/CirList/CirList.c
/*************************************************************************
 > File Name: CirList.c
 > Author:    DongZePing
 > Mail:      1939485575@qq.com
 > Time:      2019-07-27
 >
 > 1）首尾相连的循环单链表
 > 2）支持增删改查等基本操作
 ************************************************************************/
#include "CirList.h"
#include <stdio.h>
#include <stdlib.h>
//初始化
void CirList_Init(struct CirList* clist)
{
	clist->size = 0;
	clist->head = (struct Node*)malloc(sizeof(struct Node));
	clist->head->next = clist->head;
}
//尾插---修改中
void CirList_Append(struct CirList* clist, int data)
{
	struct Node* p, *temp;
	p = clist->head->next;
	while (p -> next != clist->head)
	{
		p = p->next;
	}
	temp = (struct Node*)malloc(sizeof(struct Node));
	temp->data = data;
	temp->next = clist->head;
	p->next = temp;
	clist->size++;
}
//头插
void CirList_Front(struct CirList* clist, int data)
{
	struct Node* p;
	p = (struct Node*)malloc(sizeof(struct Node));
	p->data = data;
	p->next = clist->head->next;
	clist->head->next = p;
	clist->size++;
}
//指定位置插
void CirList_Insert(struct CirList* clist, int index, int data)
{
	if (index <= 0 || index > (clist->size + 1))
	{
		return;
	}
	else {
		struct Node *p, *temp;
		p = clist->head;
		for (int i = 1; i < index; i++)
		{
			p = p->next;
		}
		temp = (struct Node*)malloc(sizeof(struct Node));
		temp->data = data;
		temp->next = p->next;
		p->next = temp;
		clist->size++;
	}
}
//指定位置删除
void CirList_Del(struct CirList* clist, int index)
{
	if (index <= 0 || index > clist->size)
	{
		return;
	}
	else {
		//先找到目标位置的前一个位置p
		struct Node *p, *temp;
		p = clist->head;
		for (int i = 1; i < index; i++)
		{
			p = p->next;
		}
		temp = p->next;
		p->next = p->next->next;
		free(temp);
		temp = NULL;
		clist->size--;
	}
}
//遍历
void CirList_Print(struct CirList* clist)
{
	struct Node *p = clist->head->next;
	while (p != clist->head)
	{
		printf("%d ", p->data);
		p = p->next;
	}
	printf("\n");
}
//销毁
void CirList_Destory(struct CirList* clist)
{
	struct Node *p, *q;
	p = clist->head;
	q = p->next;
	while (q != NULL)
	{
		free(p);
		p = q;
		q = p->next;
	}
	free(p);
	p = NULL;
}

File: Dong/Java/src/CirList/CirList.java
package CirList;

public class CirList<T> {
	private Node<T> head;
	private int size;
	//无参构造
	public CirList()
	{
		this.size = 0;
		this.head = new Node();
		this.head.next = this.head;
	}
	//尾插法
	public void CList_Append(T data)
	{
		Node<T> p, temp;
		p = this.head.next;
		while (p.next != this.head)
		{
			p = p.next;
		}
		temp = new Node<T>();
		temp.data = data;
		temp.next = this.head;
		p.next = temp;
		this.size++;
	}
	//头插法
	public void CList_Front(T data)
	{
		Node<T> p;
		p = new Node<T>();
		p.data = data;
		p.next =this.head.next;
		this.head.next = p;
		this.size++;
	}
	//随机位置插入
	public void CList_Insrt(int index, T data)
	{
		if (index <= 0 || index > (this.size + 1))
		{
			return;
		}
		else {
			Node<T> p, temp;
			p = this.head;
			for (int i = 1; i < index; i++)
			{
				p = p.next;
			}
			temp = new Node<T>();
			temp.data = data;
			temp.next = p.next;
			p.next = temp;
			this.size++;
		}
	}
	//随机位置删除
	public void CList_Del(int index)
	{
		if(index <= 0 || index > this.size)
		{
			return;
		}
		else {
			//先找到目标位置的前一个位置p
			Node<T> p, temp;
			p = this.head;
			for (int i = 1; i < index; i++)
			{
				p = p.next;
			}
			temp = p.next;
			p.next = p.next.next;
			temp = null;
			this.size--;
		}
	}
	//输出
	public void CList_Print()
	{

		Node<T> p = this.head.next;
		while (p != this.head)
		{
			System.out.print(p.data+" ");
			p = p.next;
		}
	System.out.println();
	}	
}